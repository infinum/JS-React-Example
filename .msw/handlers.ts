import { isJsonApiClass, isModelPersisted, modelToJsonApi } from '@datx/jsonapi';
import { buildMany } from '@datx/test-data-factory';
import { rest } from 'msw';
import { sessionFactory } from '../__factories__/session';
import { userFactory } from '../__factories__/user';
import { flightFactory } from '../__factories__/flight';
import { Session } from '@/models/Session';
import { User } from '@/models/User';

export const handlers = [
	// datx.get({
	// 	op: 'getMany',
	// 	type: Flight.type,
	// } as const, (req, res, ctx) => {

	// }),
	rest.get(`${process.env.NEXT_PUBLIC_API_ENDPOINT}/sessions/current`, (req, res, ctx) => {
		userFactory.reset();
		sessionFactory.reset();
		const include = req.url.searchParams.get('include');
		const session = sessionFactory();

		const included = include
			?.split(',')
			.map((include) => {
				const includeData = session[include];

				if (Array.isArray(includeData)) {
					return includeData.map((item) => modelToJsonApi(item));
				}

				const rawModel = modelToJsonApi(includeData);

				rawModel.id = includeData.id;

				return rawModel;
			})
			.flat();

		const rawSessionModel = modelToJsonApi(session);
		rawSessionModel.id = session.id;

		return res(
			ctx.json({
				data: rawSessionModel,
				included,
			})
		);
	}),
	rest.get(`${process.env.NEXT_PUBLIC_API_ENDPOINT}/flights`, (_req, res, ctx) => {
		flightFactory.reset();
		const data = buildMany(flightFactory, 2);

		return res(
			ctx.json({
				data: data.map((flight) => modelToJsonApi(flight)),
			})
		);
	}),
];
