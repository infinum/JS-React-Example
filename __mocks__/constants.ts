export const TEST_INVALID_EMAIL = 'Email User not found';

export const TEST_INVALID_PASSWORD = 'Password Invalid password';

export const TEST_FLIGHT_NAME = 'Password Invalid password';
