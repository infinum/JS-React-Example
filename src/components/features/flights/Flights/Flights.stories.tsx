import { ComponentMeta, ComponentStory } from '@storybook/react';

import { Flights } from './Flights';

export default {
	title: 'UI/features/Flights',
	component: Flights,
} as ComponentMeta<typeof Flights>;

const Template: ComponentStory<typeof Flights> = (args) => <Flights {...args} />;

export const Basic = Template.bind({});
Basic.args = {
	/* the args you need here will depend on your component */
};
