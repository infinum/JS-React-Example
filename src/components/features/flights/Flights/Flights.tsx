import { FC, Fragment, useMemo } from 'react';
import { useSession } from '@/hooks/use-session';
import { useDatx } from '@datx/swr';
import { getFlightsQuery } from '@/queries/flights';
import { EmptyListMessage } from '@/components/shared/messages/EmptyListMessage/EmptyListMessage';
import { FlightList, FlightListFallback } from '@/components/shared/flight/FlightList/FlightList';
import { BasicPagination, getPagination } from '@/components/shared/paginations/BasicPagination/BasicPagination';
import { useRouter } from 'next/router';

export interface IFlightsProps {}

export const Flights: FC<IFlightsProps> = () => {
	const { query } = useRouter();
	const { data: sessionResponse } = useSession();
	const user = sessionResponse?.data.user;
	const { data, error } = useDatx(() => getFlightsQuery(user, query.page as string));
	const pagination = useMemo(() => getPagination(data), [data]);

	if (error) {
		return <div>error</div>;
	}

	if (!data) {
		return <FlightListFallback />;
	}

	if (data.data.length === 0) {
		return <EmptyListMessage />;
	}

	return (
		<Fragment>
			<FlightList flightList={data.data} />
			<BasicPagination pagination={pagination} />
		</Fragment>
	);
};
