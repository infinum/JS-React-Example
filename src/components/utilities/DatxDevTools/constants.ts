export const separator = '@';
export const errorKeyIndicator = 'err';
export const isValidatingKeyIndicator = 'validating';
export const manyKeyIndicator = 'many';
