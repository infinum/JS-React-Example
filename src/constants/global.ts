export const apiEndpoint = process.env.NEXT_PUBLIC_API_ENDPOINT;
