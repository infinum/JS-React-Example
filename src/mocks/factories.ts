import { createClient } from '@/datx/create-client';
import { Session } from '@/models/Session';
import { User } from '@/models/User';
import { createFactory, oneOf, perBuild, sequence } from '@datx/test-data-factory';

const client = createClient();
const factory = createFactory(client);

export const userFactory = factory(User, {
	fields: {
		id: sequence(),
		firstName: oneOf('Johnny', 'Michael', 'Philip', 'Darius', 'Nieves', 'Christian'),
		lastName: oneOf('Smith', 'Brown', 'Miller'),
	},
});

export const sessionFactory = factory(Session, {
	fields: {
		id: sequence(),
		email: sequence((x) => `user${x}@example.com`),
		user: perBuild(() => userFactory()),
	},
});
