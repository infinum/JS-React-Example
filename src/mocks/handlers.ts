import { buildJsonApiDocument } from '@/datx/buildJsonApiDocument';
import { TEST_INVALID_EMAIL, TEST_INVALID_PASSWORD } from '__mocks__/constants';
import { sessionFactory } from '__mocks__/factories';
import { rest } from 'msw';

export const generateTestApiUrl = (path: string) => {
	return process.env.API_TEST_ENDPOINT + path;
};

export const MOCKED_URLS = {
	SessionCurrent: generateTestApiUrl('sessions/current'),
	SessionMe: generateTestApiUrl('sessions/me'),
	Session: generateTestApiUrl('sessions'),
	Flights: generateTestApiUrl('flights'),
} as const;

export const handlers = [
	rest.get(MOCKED_URLS.SessionCurrent, async (_req, res, ctx) => {
		return res(
			ctx.status(401),
			ctx.json({
				errors: [
					{
						status: 'unauthorized',
						code: 'UNAUTHORIZED',
						title: 'Unauthorized',
						detail: 'Must be logged in to perform this action',
					},
				],
			})
		);
	}),
	rest.post(MOCKED_URLS.Session, async (req, res, ctx) => {
		const body = await req.json();
		const emailOverride = body.data?.attributes?.email || undefined;

		return res(
			ctx.status(200),
			ctx.json(
				sessionFactory({
					overrides: {
						email: emailOverride,
					},
				})
			)
		);
	}),
	rest.delete(MOCKED_URLS.SessionMe, async (_req, res, ctx) => {
		return res(ctx.status(204));
	}),
];

export const handlerOverrides = {
	invalidEmailLogin: rest.post(MOCKED_URLS.Session, (_req, res, ctx) => {
		return res(
			ctx.status(422),
			ctx.json({
				errors: [
					{
						status: 'unprocessable entity',
						code: 'UNPROCESSABLE_ENTITY',
						title: 'Unprocessable Entity',
						detail: TEST_INVALID_EMAIL,
						source: { parameter: 'email', pointer: 'data/attributes/email' },
					},
				],
			})
		);
	}),
	invalidPasswordLogin: rest.post(MOCKED_URLS.Session, (_req, res, ctx) => {
		return res(
			ctx.status(422),
			ctx.json({
				errors: [
					{
						status: 'unprocessable entity',
						code: 'UNPROCESSABLE_ENTITY',
						title: 'Unprocessable Entity',
						detail: TEST_INVALID_PASSWORD,
						source: { parameter: 'password', pointer: 'data/attributes/password' },
					},
				],
			})
		);
	}),
	activeCurrentSession: rest.get(MOCKED_URLS.SessionCurrent, async (req, res, ctx) => {
		const session = sessionFactory();

		return res(ctx.status(200), ctx.json(buildJsonApiDocument(session)));
	}),
};
