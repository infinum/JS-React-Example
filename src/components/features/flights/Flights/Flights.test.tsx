import { axe } from 'jest-axe';
import { render, screen, waitForElementToBeRemoved } from 'test-utils';
import { Flights } from './Flights';
import { server } from '@/mocks/server';
import { rest } from 'msw';
import { MOCKED_URLS, handlerOverrides } from '@/mocks/handlers';
import { flightsFactoryJsonApi } from '__mocks__/factories';
import { TEST_FLIGHT_NAME } from '__mocks__/constants';

describe('Flights', () => {
	beforeEach(() => {
		server.use(handlerOverrides.activeCurrentSession);
	});

	it('should be accessible', async () => {
		server.use(
			rest.get(MOCKED_URLS.Flights, (req, res, ctx) => {
				return res(ctx.status(200), ctx.json({ data: [] }));
			})
		);

		const { container } = render(<Flights />);

		await waitForElementToBeRemoved(() => screen.queryByText('Loading...'));

		const results = await axe(container);

		expect(results).toHaveNoViolations();
	});

	it('should render loading state', () => {
		server.use(
			rest.get(MOCKED_URLS.Flights, (req, res, ctx) => {
				return res(ctx.status(200), ctx.json(flightsFactoryJsonApi(5)));
			})
		);

		render(<Flights />);

		expect(screen.getByText('Loading...')).toBeInTheDocument();
	});

	it('should render error state', async () => {
		server.use(
			rest.get(MOCKED_URLS.Flights, (req, res, ctx) => {
				return res(
					ctx.status(404),
					ctx.json({
						error: 'Error',
					})
				);
			})
		);

		render(<Flights />);

		await waitForElementToBeRemoved(() => screen.queryByText('Loading...'));

		expect(screen.getByText('error')).toBeInTheDocument();
	});

	it('should render 10 flights', async () => {
		server.use(
			rest.get(MOCKED_URLS.Flights, (req, res, ctx) => {
				const data = flightsFactoryJsonApi(10);

				return res(
					ctx.status(200),
					ctx.json({
						data,
					})
				);
			})
		);

		render(<Flights />);

		await waitForElementToBeRemoved(() => screen.queryByText('Loading...'));

		expect(screen.getAllByRole('heading', { name: TEST_FLIGHT_NAME })).toHaveLength(10);
	});
});
