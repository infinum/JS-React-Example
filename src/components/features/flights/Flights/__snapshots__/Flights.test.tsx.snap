// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flights should render error state 1`] = `
<DocumentFragment>
  @keyframes animation-0 {
  from {
    border-color: var(--skeleton-start-color);
    background: var(--skeleton-start-color);
  }

  to {
    border-color: var(--skeleton-end-color);
    background: var(--skeleton-end-color);
  }
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*:not(style)~*:not(style) {
  margin-top: var(--chakra-space-4);
  -webkit-margin-end: 0px;
  margin-inline-end: 0px;
  margin-bottom: 0px;
  -webkit-margin-start: 0px;
  margin-inline-start: 0px;
}

.emotion-1 {
  --skeleton-start-color: var(--chakra-colors-gray-100);
  --skeleton-end-color: var(--chakra-colors-gray-400);
  background: var(--skeleton-start-color);
  border-color: var(--skeleton-end-color);
  opacity: 0.7;
  border-radius: var(--chakra-radii-sm);
  -webkit-animation: 0.8s linear infinite alternate animation-0;
  animation: 0.8s linear infinite alternate animation-0;
  box-shadow: var(--chakra-shadows-none);
  -webkit-background-clip: padding-box;
  background-clip: padding-box;
  cursor: default;
  color: var(--chakra-colors-transparent);
  pointer-events: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: var(--chakra-sizes-full);
  height: 100px;
  padding: var(--chakra-space-5);
}

.chakra-ui-dark .emotion-1:not([data-theme]),
[data-theme=dark] .emotion-1:not([data-theme]),
.emotion-1[data-theme=dark] {
  --skeleton-start-color: var(--chakra-colors-gray-800);
  --skeleton-end-color: var(--chakra-colors-gray-600);
}

.emotion-1::before,
.emotion-1::after,
.emotion-1 * {
  visibility: hidden;
}

<div
    class="chakra-stack emotion-0"
  >
    <span>
      Loading...
    </span>
    <div
      class="chakra-skeleton emotion-1"
    />
    <div
      class="chakra-skeleton emotion-1"
    />
    <div
      class="chakra-skeleton emotion-1"
    />
    <div
      class="chakra-skeleton emotion-1"
    />
    <div
      class="chakra-skeleton emotion-1"
    />
    <div
      class="chakra-skeleton emotion-1"
    />
    <div
      class="chakra-skeleton emotion-1"
    />
    <div
      class="chakra-skeleton emotion-1"
    />
    <div
      class="chakra-skeleton emotion-1"
    />
    <div
      class="chakra-skeleton emotion-1"
    />
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
